File: java/com/countrygamer/arcanacraft/common/KeyHandler.java
package com.countrygamer.arcanacraft.common;

import java.util.ArrayList;

import net.minecraft.client.Minecraft;
import net.minecraft.client.settings.KeyBinding;

import org.lwjgl.input.Keyboard;

import com.countrygamer.arcanacraft.common.network.MessageCastQuom;
import com.countrygamer.arcanacraft.common.network.MessageSelectQuom;
import com.countrygamer.countrygamercore.Base.common.network.PacketHandler;

import cpw.mods.fml.client.registry.ClientRegistry;
import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.common.gameevent.InputEvent.KeyInputEvent;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class KeyHandler {
	
	private static final ArrayList<String> descriptions = new ArrayList<String>();
	
	private static final int[] keyValues = {
			Keyboard.KEY_P, Keyboard.KEY_COMMA, Keyboard.KEY_PERIOD
	};
	
	private final KeyBinding[] keys;
	
	public KeyHandler() {
		descriptions.add("key.cast.desc");
		descriptions.add("key.lastQuom.desc");
		descriptions.add("key.nextQuom.desc");
		
		this.keys = new KeyBinding[descriptions.size()];
		
		for (int i = 0; i < descriptions.size(); i++) {
			this.keys[i] = new KeyBinding(descriptions.get(i), keyValues[i],
					"keys.arcanacraft.category");
			ClientRegistry.registerKeyBinding(this.keys[i]);
		}
	}
	
	@SideOnly(Side.CLIENT)
	@SubscribeEvent
	public void onKeyInput(KeyInputEvent event) {
		if (!Minecraft.getMinecraft().inGameHasFocus) return;
		
		KeyBinding cast = this.getKey("key.cast.desc");
		KeyBinding lastQuom = this.getKey("key.lastQuom.desc");
		KeyBinding nextQuom = this.getKey("key.nextQuom.desc");
		
		if (cast.isPressed()) {
			// System.out.println("Press Key");
			// PacketCastQuom packet = new PacketCastQuom();
			// ArcanaCraft.instance.packetChannel.sendToServer(packet);
			PacketHandler.sendToAll(ArcanaCraft.pluginID, new MessageCastQuom());
			PacketHandler.sendToServer(ArcanaCraft.pluginID, new MessageCastQuom());
		}
		
		if (lastQuom.isPressed()) {
			// PacketSelectQuom packet = new PacketSelectQuom(false);
			// ArcanaCraft.instance.packetChannel.sendToServer(packet);
			PacketHandler.sendToServer(ArcanaCraft.pluginID, new MessageSelectQuom(false));
		}
		
		if (nextQuom.isPressed()) {
			// PacketSelectQuom packet = new PacketSelectQuom(true);
			// ArcanaCraft.instance.packetChannel.sendToServer(packet);
			PacketHandler.sendToServer(ArcanaCraft.pluginID, new MessageSelectQuom(true));
		}
		
	}
	
	public KeyBinding getKey(String desc) {
		return this.keys[descriptions.indexOf(desc)];
	}
	
}

File: src/main/scala/com/countrygamer/arcanacraft/client/KeyHandler.scala
package com.countrygamer.arcanacraft.client

import java.util

import com.countrygamer.arcanacraft.common.ArcanaCraft
import com.countrygamer.arcanacraft.common.network.PacketKeyPress
import com.countrygamer.cgo.common.network.PacketHandler
import cpw.mods.fml.client.registry.ClientRegistry
import cpw.mods.fml.common.eventhandler.SubscribeEvent
import cpw.mods.fml.common.gameevent.InputEvent.KeyInputEvent
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.Minecraft
import net.minecraft.client.settings.KeyBinding
import org.lwjgl.input.Keyboard

/**
 *
 *
 * @author CountryGamer
 */
@SideOnly(Side.CLIENT)
object KeyHandler {

	private final val keyValues: util.HashMap[String, Int] = new util.HashMap[String, Int]()

	private var keys: util.HashMap[String, KeyBinding] = null

	{
		keyValues.put("key.cast.desc", Keyboard.KEY_P)
		keyValues.put("key.lastQuom.desc", Keyboard.KEY_COMMA)
		keyValues.put("key.nextQuom.desc", Keyboard.KEY_PERIOD)

		this.keys = new util.HashMap[String, KeyBinding]()
		val iterator: util.Iterator[String] = this.keyValues.keySet().iterator()
		while (iterator.hasNext) {
			val desc: String = iterator.next()
			val key: Int = this.keyValues.get(desc)
			this.keys.put(desc, new KeyBinding(desc, key, "keys.arcanacraft.category"))
			ClientRegistry.registerKeyBinding(this.keys.get(desc))
		}

	}

	@SubscribeEvent
	def onKeyInput(event: KeyInputEvent): Unit = {
		if (!Minecraft.getMinecraft.inGameHasFocus) return

		val cast: KeyBinding = this.getKey("key.cast.desc")
		val lastQuom: KeyBinding = this.getKey("key.lastQuom.desc")
		val nextQuom: KeyBinding = this.getKey("key.nextQuom.desc")

		var packet: PacketKeyPress = null
		if (cast != null && cast.isPressed) {
			packet = new PacketKeyPress(0)
		}
		else if (lastQuom != null && lastQuom.isPressed) {
			packet = new PacketKeyPress(1)
		}
		else if (nextQuom != null && nextQuom.isPressed) {
			packet = new PacketKeyPress(2)
		}

		if (packet != null) {
			PacketHandler.sendToServer(ArcanaCraft.pluginID, packet)
			// TODO Does this work? (sending to client)
			PacketHandler
					.sendToPlayer(ArcanaCraft.pluginID, packet, Minecraft.getMinecraft.thePlayer)
		}
	}

	def getKey(desc: String): KeyBinding = {
		if (this.keys.containsKey(desc)) {
			return this.keys.get(desc)
		}
		null
	}

}
